contracts/erc721/ERC721PreApproval.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2021 the ethier authors (github.com/divergencetech/ethier)
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../thirdparty/opensea/OpenSeaGasFreeListing.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+/// @notice Pre-approval of OpenSea proxies for gas-less listing
+/// @dev This wrapper allows users to revoke the pre-approval of their
+/// associated proxy and emits the corresponding events. This is necessary for
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
+    /// @dev It is important that Active remains at first position, since this
+    /// is the scenario that we are trying to optimize for.
```

Comment:
Agreed, will add a test.

Yes, this allows us to optimize for the `Active` case because no overhead is needed to store it.

Reply:
If I understood correctly, this is not testable from the outside because the state is never exported - so `abigen` will not translate it. Maybe we should effectively lock this by testing transaction costs?

Reply:
On second thought, the default behaviour would break if `Active` was non-zero, correct? If so, I guess we can leave it.

If you think we should have an explicit test, I would add a `TestableERC721PreApproval.sol` with functions `ActiveState` and `InactiveState`

Reply:
Yes, it would break in that case. So I'm opting to not check this explicitly for now.